### PR TITLE
CFE-3433: action_policy and dry-run handling with custom promise modules

### PR DESCRIPTION
**Pull request by vpodzime**

### CODE REVIEW

File: libpromises/mod_custom.c [hunk 1]
```diff
@@ -325,7 +334,8 @@ static JsonElement *PromiseModule_Receive(PromiseModule *module, const Promise *
             for (size_t i = 0; i < length; ++i)
             {
                 logged_error |= PromiseModule_LogJson(
-                    JsonArrayGet(json_log_messages, i), pp, promise_log_level);
+                    JsonArrayGet(json_log_messages, i), pp,
+                    promise_log_level, changes_allowed);
```
Comment (olehermanse) [inline libpromises/mod_custom.c:338]: Reading these binpacked / Z shape lists of arguments is really annoying to me, but maybe just what you are used to.

File: libpromises/mod_custom.c [hunk 2]
```diff
@@ -125,13 +125,21 @@ static bool GetInterpreterAndPath(
...
 {
     const char *level_string = JsonObjectGetAsString(object, "level");
     const char *message = JsonObjectGetAsString(object, "message");
 
     assert(level_string != NULL && message != NULL);
     const LogLevel level = LogLevelFromString(level_string);
+
+    if ((level == LOG_LEVEL_INFO) && !changes_allowed)
+    {
+        Log(LOG_LEVEL_NOTICE,
```
[the remark this answers is not in the document]
Comment (nickanderson) [reply libpromises/mod_custom.c:139]: My first instinct is error

File: libpromises/mod_custom.c [hunk 3]
```diff
@@ -899,7 +909,7 @@ static bool PromiseModule_Validate(PromiseModule *module, const EvalContext *ctx
     PromiseModule_Send(module);
 
     // Prints errors / log messages from module:
-    JsonElement *response = PromiseModule_Receive(module, pp);
+    JsonElement *response = PromiseModule_Receive(module, pp, !dontdo /* changes_allowed */);
```
Comment (olehermanse) [inline libpromises/mod_custom.c:912]: Since this is in the validate operation, maybe we should send `changes_allowed = false`?

**Comment by vpodzime** [comment]
@cf-bottom run this through jenkins, please, so that we see where it fails

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/8424/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8424/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8424/

**Comment by vpodzime** [comment]
@cf-bottom test this in jenkins with exotics, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/8447/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8447/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8447/

**Comment by vpodzime** [comment]
> Sure, I triggered a build:
> 
> [Build Status](https://ci.cfengine.com//job/pr-pipeline/8447/)
> 
> (with exotics)

Sigh... Didn't even get to testing my changes here on exotics, all failed earlier.

**Comment by vpodzime** [comment]
Exotics only: [Build Status](https://ci.cfengine.com/job/pr-pipeline/8479/)

**Comment by vpodzime** [comment]
> Exotics only: [Build Status](https://ci.cfengine.com/job/pr-pipeline/8479/)

Passed on AIX 5.3 and Solaris 11, the only two platforms where it had a chance to succeed.

**Review by larsewi — APPROVED** [review]
ACK

**Review by olehermanse — APPROVED** [review]
LGTM, thanks 👍

File: libpromises/mod_custom.c [hunk 4]
```diff
@@ -175,7 +167,8 @@ static inline JsonElement *PromiseModule_ParseResultClasses(char *value)
     return result_classes;
 }
 
-static JsonElement *PromiseModule_Receive(PromiseModule *module, const Promise *pp)
+static JsonElement *PromiseModule_Receive(PromiseModule *module, const Promise *pp,
+                                          uint16_t n_log_msgs[LOG_LEVEL_DEBUG + 1])
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/mod_custom.c:171]: Don't be lazy :wink:

**Pull request merged**